README.md
Alephium Mining Companion
====

This project is a companion for Alephium full nodes, providing some convenience features
when enabling mining:

1. Creates a mining wallet if none exists. 
   It can also restore the wallet if mnemonics are provided.
2. Configures the mining process to use the created wallet, 
   if one created or different from the one configured.
3. Starts the mining automatically when the node is in sync
4. Transfers the mining reward to another provided wallet/address
   on a fixed frequency (hourly, daily, ...). Note that the wallet
   can be created anywhere as long as it's a valid Alephium address
5. Expose metrics for both the mining and the transfer wallets

Companion means it needs to run alongside an Alephium full node,
and particularly have REST connectivity to it (default port 12973)

![mining-sidecar](alephium-mining.png)

# Configuration

## Docker compose

Docker-compose is a good way of running the side, assuming the
[Alephium node is also running in docker](https://touille.io/posts/how-to-run-alephium-full-node/).

```
version: "3"
services:
  mining-companion:
    image: touilleio/alephium-mining-companion:v6
    restart: unless-stopped
    security_opt:
      - no-new-privileges:true
    environment:
      - ALEPHIUM_ENDPOINT=http://broker:12973
      - LOG_LEVEL=debug
      - WALLET_NAME=$miningWalletName
      - WALLET_PASSWORD=$miningWalletPassword
      - WALLET_MNEMONIC=$miningWalletMnemonic
      - TRANSFER_ADDRESS=$miningTransferAddress
      - TRANSFER_FREQUENCY=5m
      - TRANSFER_MIN_AMOUNT=50000000000000000000
      - IMMEDIATE_TRANSFER=true
      - START_MINING=false
```
with variables `miningWalletName`, `miningWalletPassword`, `miningWalletMnemonic` and `miningTransferAddress`
in a `.env` file in the same folder as the `docker-compose.yml`.

| Variable | Default | Description |
|----------|---------|-------------|
| `ALEPHIUM_ENDPOINT` | `http://alephium:12973` | REST URI of your Alephium node. Mind localhost in a docker container point to the docker container, not the host itself. |
| `ALEPHIUM_API_KEY` | _optional_ | API key to use to connect to `ALEPHIUM_ENDPOINT`. |
| `WALLET_NAME` | `mining-companion-wallet-1` | Name of the miner wallet |
| `WALLET_PASSWORD` | `Default-Password-1234` | Password to unlock the miner wallet |
| `WALLET_MNEMONIC` | _optional_ | Mnemonic to restore (create) the wallet if it does not exist. Random mnemonic will be generated if not set |
| `WALLET_MNEMONIC_PASSPHRASE` | _optional_ | A passphrase associated with the mnemonic, if any |
| `TRANSFER_MIN_AMOUNT` | 20000000000000000000 (20 ALF) | Min amount to transfer at once. It uses the `sweepAll` function to optimize the transaction. |
| `TRANSFER_ADDRESS` | _optional_ | Address to transfer the mining rewards to. If none provided, no transfer is performed. Double check you're sending the funds to the right address !! |
| `TRANSFER_FREQUENCY` | `15m` | Frequency at which funds are transferred |
| `PRINT_MNEMONIC` | `true` | If true and a wallet is created without pre-set mnemonic (`WALLET_MNEMONIC` option above), the randomly generate mnemonic is printed out. This is a sensitive information, use it with caution! |
| `IMMEDIATE_TRANSFER` | `false` | If set to true, a transfer is sent at the start of the container, without waiting for `TRANSFER_FREQUENCY` initial time |
| `START_MINING` | `false` | If set to true, the mining machinery built-in the broker will start mining. This is disabled by default and the dedicated, more efficient [CPU miner](https://github.com/alephium/cpu-miner) is recommended for mining as the time of writing |

## Docker

Replace `123456789012345678901234567890123456789012345` below with your own wallet address!

```
docker run -it --rm --link alephium:alephium -e TRANSFER_ADDRESS=123456789012345678901234567890123456789012345 touilleio/alephium-mining-companion:v6
```

As a reminder, running a Alephium full node looks like the following:

```
docker run -it --rm --name alephium -p 12973:12973 alephium/alephium:v1.1.7
```


CHANGELOG.md
Alephium Mining Companion Changelog
====

# Version v7.1.2

- Bump alephium/go-sdk to v1.6.3

# Version v7.1.1

- Bump alephium/go-sdk to v1.6.2

# Version v7.1.0

- Bump alephium/go-sdk to v1.6

# Version v7.0.0

- Migrate to github.com/alephium/go-sdk
- Use sweepAll function

# Version v6.1.0

## Improvements

- Add support for API key via `ALEPHIUM_API_KEY` environment variable
- Remove no longer needed `START_MINING` environment variable and feature

# Version v6.0.0

## Breaking behaviour changes

- Use `SweepAll` function instead of Transfer, once the `TRANSFER_MIN_AMOUNT` threshold is met per address
- `TRANSFER_MAX_AMOUNT` is no longer used.
- `TRANSFER_MIN_AMOUNT` default changed to 20 ALPH

## Improvements

- Use `alephium.GetAddressesAsString`

# Version v5.1.1

## Improvement changes

- Add markers around the mnemonic to make it more visible

# Version v5.1.0

## Improvement changes

- Add metrics to monitor balance of miner and transfer addresses

# Version v5.0.0

## Breaking changes

- Follow up changes in v1.0.0 (mainnet)

# Version v4.0.0

## Breaking changes

- Follow up changes in v0.11.0

# Version v3.0.0

## Behavior changes

- START_MINING has be set to `false` by default, meaning the mining-companion is setting the mining addresses,
  but does not start the full node mining, since this feature is deprecated in favor of the 
  [CPU miner](https://github.com/alephium/cpu-miner)

# Version v2.1.0

## Improvements

- Expose metrics of number of amount transferred

# Version v2.0.0

- Rename mining-sidecar to mining-companion

# Version v1.5.1

## Improvements

- Follow upstream update in alephium-go-client to simplify code

# Version v1.5.0

## Improvements

- Adds `START_MINING` option, telling whether to start local mining. Useful when running external miner (brought by v0.8.5)

# Version v1.4.0

## Improvements

- Follow up API name change of v0.8.2

# Version v1.3.0

## Fix

- Followup change in coinbase lockup time

# Version v1.2.1

## Fix

- Followup change in miner addresses behaviour

# Version v1.2.0

## Improvements

- Add `TRANSFER_MIN_AMOUNT` configuration option
- Bump alephium-go-client dependency and refactor to use AFL manipulation module

# Version v1.1.0

## Improvements

- Add `IMMEDIATE_TRANSFER` configuration option
- Bump alephium-go-client dependency
- Enhance logging messages

# Version v1.0.0

- Initial version


alph_test.go
package main

import (
	"encoding/json"
	"fmt"
	"github.com/stretchr/testify/assert"
	"math/rand"
	"os"
	"testing"
	"time"
)

func TestALPHConstruct(t *testing.T) {
	a1, ok := ALPHFromALPHString("12")
	assert.True(t, ok)
	assert.Equal(t, 12.000000000, a1.FloatALPH())
	assert.Equal(t, fmt.Sprintf("12.000000000%s", N), a1.PrettyString())
	assert.Equal(t, "12000000000000000000", a1.String())

	a2, ok := ALPHFromCoinString("12")
	assert.True(t, ok)
	assert.Equal(t, "12", a2.PrettyString())
	assert.Equal(t, "12", a2.String())
}

func TestZeroALPH(t *testing.T) {
	a1 := ALPH{}
	assert.Equal(t, "0", a1.String())
}

func TestALPHConstructWithDot(t *testing.T) {
	a1, ok := ALPHFromALPHString("12.12")
	assert.True(t, ok)
	assert.Equal(t, 12.120000000, a1.FloatALPH())
}

func TestALPHAdd(t *testing.T) {
	a1, ok := ALPHFromALPHString("10.1")
	assert.True(t, ok)
	a2, ok := ALPHFromALPHString("2.02")
	assert.True(t, ok)
	a3, ok := ALPHFromALPHString("12.12")
	assert.True(t, ok)

	res := a1.Add(a2)
	assert.Equal(t, 0, res.Cmp(a3))
	assert.Equal(t, res, a3)
}

func TestALPHSub(t *testing.T) {
	a1, ok := ALPHFromALPHString("10")
	assert.True(t, ok)
	a2, ok := ALPHFromALPHString("2")
	assert.True(t, ok)
	a3, ok := ALPHFromALPHString("12")
	assert.True(t, ok)

	res := a3.Subtract(a2)
	assert.Equal(t, 0, res.Cmp(a1))
	assert.Equal(t, res, a1)
}

func TestALPHMul(t *testing.T) {
	a1, ok := ALPHFromALPHString("10")
	assert.True(t, ok)
	a2, ok := ALPHFromALPHString("2")
	assert.True(t, ok)
	a3, ok := ALPHFromALPHString("12")
	assert.True(t, ok)

	res := a2.Multiply(5)
	assert.Equal(t, 0, res.Cmp(a1))
	assert.Equal(t, res, a1)
	res = a2.Multiply(6)
	assert.Equal(t, 0, res.Cmp(a3))
	assert.Equal(t, res, a3)
}

func TestALPHDiv(t *testing.T) {
	a1, ok := ALPHFromALPHString("10")
	assert.True(t, ok)
	a2, ok := ALPHFromALPHString("2")
	assert.True(t, ok)
	a3, ok := ALPHFromALPHString("12")
	assert.True(t, ok)

	res := a1.Divide(5)
	assert.Equal(t, 0, res.Cmp(a2))
	assert.Equal(t, res, a2)
	res = a3.Divide(6)
	assert.Equal(t, 0, res.Cmp(a2))
	assert.Equal(t, res, a2)
}

func TestALPHDecode(t *testing.T) {
	a1, ok := ALPHFromALPHString("10")
	assert.True(t, ok)
	a2, ok := ALPHFromALPHString("2")
	assert.True(t, ok)
	a3, ok := ALPHFromALPHString("12")
	assert.True(t, ok)

	res := a1.Divide(5)
	assert.Equal(t, 0, res.Cmp(a2))
	assert.Equal(t, res, a2)
	res = a3.Divide(6)
	assert.Equal(t, 0, res.Cmp(a2))
	assert.Equal(t, res, a2)
}

func TestRoundAmount(t *testing.T) {
	rand.Seed(time.Now().UnixNano() + int64(os.Getpid()))
	alph := RandomALPHAmount(100)
	fmt.Printf("%s", alph.PrettyString())

	TenALPH, ok := ALPHFromALPHString("10")
	assert.True(t, ok)

	// just to ensure at least 10 ALPH
	alph = alph.Add(TenALPH)
	assert.True(t, alph.Cmp(TenALPH) > 0)
	var finalTxAmount ALPH
	fiveALPH := TenALPH.Divide(2)
	txAmount := alph.Subtract(fiveALPH)
	if txAmount.Cmp(TenALPH) > 0 {
		finalTxAmount = TenALPH
	} else {
		finalTxAmount = txAmount
	}

	assert.True(t, finalTxAmount.Cmp(TenALPH) <= 0)
}

func TestNanoAmount(t *testing.T) {
	rand.Seed(time.Now().UnixNano() + int64(os.Getpid()))
	alph := RandomNanoALPHAmount(int(OneBillionInt64))
	fmt.Printf("%s\n", alph.PrettyString())
	fmt.Printf("%.4f\n", alph.FloatALPH())

	alph = RandomNanoALPHAmount(10000000)
	fmt.Printf("%s\n", alph.PrettyString())
}

type TestJsonStruct struct {
	Amount ALPH `json:"Amount"`
}

func TestJSON(t *testing.T) {
	rand.Seed(time.Now().UnixNano() + int64(os.Getpid()))
	alph := RandomNanoALPHAmount(int(OneBillionInt64))

	j1 := TestJsonStruct{
		Amount: alph,
	}

	b, err := json.Marshal(j1)
	assert.Nil(t, err)

	fmt.Printf("-->%s\n", string(b))

	j2 := &TestJsonStruct{}
	err = json.Unmarshal(b, j2)
	assert.Nil(t, err)

	fmt.Printf("%s\n", j2.Amount.PrettyString())
}


alph.go
package main

import (
	"bytes"
	"fmt"
	"github.com/willf/pad"
	"math/big"
	"math/rand"
	"strings"
)

type ALPH struct {
	Amount *big.Int
}

var (
	OneQuintillionString = "1000000000000000000"
	OneQuintillionInt64  = int64(1000000000000000000)
	OneBillionInt64      = int64(1000000000)
	CoinInOneALPH        = new(big.Int).SetInt64(OneQuintillionInt64)
	CoinInNanoALPH       = new(big.Int).SetInt64(OneBillionInt64)
	//N = "א"
	N = "ALPH"
)

func ALPHFromALPHString(amount string) (ALPH, bool) {
	split := strings.Split(amount, ".")
	if len(split) == 1 {
		alphAmount, ok := new(big.Int).SetString(amount, 10)
		if ok {
			coinAmount := new(big.Int).Mul(alphAmount, CoinInOneALPH)
			return ALPH{Amount: coinAmount}, true
		}
	} else if len(split) == 2 {
		decimals := pad.Right(split[1], 18, "0")
		coinAmount, ok := new(big.Int).SetString(strings.Join([]string{split[0], decimals}, ""), 10)
		if ok {
			return ALPH{Amount: coinAmount}, true
		}
	}
	return ALPH{}, false
}

func ALPHFromCoinString(amount string) (ALPH, bool) {
	alphAmount, ok := new(big.Int).SetString(amount, 10)
	if !ok {
		return ALPH{}, false
	}
	return ALPH{Amount: alphAmount}, true
}

func (alph ALPH) Add(other ALPH) ALPH {
	c := new(big.Int)
	c.Add(alph.Amount, other.Amount)
	return ALPH{Amount: c}
}

func (alph ALPH) Subtract(other ALPH) ALPH {
	c := new(big.Int)
	c.Sub(alph.Amount, other.Amount)
	return ALPH{Amount: c}
}

func (alph ALPH) Multiply(multiplier int64) ALPH {
	c := new(big.Int)
	m := new(big.Int).SetInt64(multiplier)
	c.Mul(alph.Amount, m)
	return ALPH{Amount: c}
}

func (alph ALPH) Divide(divider int64) ALPH {
	c := new(big.Int)
	m := new(big.Int).SetInt64(divider)
	c.Div(alph.Amount, m)
	return ALPH{Amount: c}
}

func (alph ALPH) Cmp(other ALPH) int {
	return alph.Amount.Cmp(other.Amount)
}

func (alph ALPH) String() string {
	if alph.Amount == nil {
		return "0"
	}
	return alph.Amount.String()
}

func (alph ALPH) PrettyString() string {
	if alph.Amount == nil {
		return "0"
	}
	if alph.Amount.Cmp(CoinInNanoALPH) > 0 {
		return fmt.Sprintf("%s%s", strings.TrimRight(strings.TrimRight(fmt.Sprintf("%.9f", alph.FloatALPH()), "0"), "."), N)
	}
	return alph.Amount.String()
}

func (alph ALPH) FloatALPH() float64 {
	c := new(big.Int)
	nanoAFL := c.Div(alph.Amount, CoinInNanoALPH).Int64()
	return float64(nanoAFL) / float64(OneBillionInt64)
}

func RandomALPHAmount(upperLimit int) ALPH {
	unit := rand.Intn(upperLimit)
	decimals := rand.Intn(int(OneBillionInt64))
	rAmountStr := fmt.Sprintf("%d.%d", unit, decimals)
	alph, _ := ALPHFromALPHString(rAmountStr)
	return alph
}

func RandomNanoALPHAmount(upperLimit int) ALPH {
	nanoALPH := rand.Intn(upperLimit)
	c := new(big.Int).SetInt64(int64(nanoALPH))
	m := new(big.Int).Mul(c, CoinInNanoALPH)
	return ALPH{Amount: m}
}

func (alph ALPH) MarshalJSON() ([]byte, error) {
	buffer := bytes.NewBufferString(`"`)
	str := alph.Amount.String()
	buffer.WriteString(str)
	buffer.WriteString(`"`)
	return buffer.Bytes(), nil
}

func (alph *ALPH) UnmarshalJSON(b []byte) error {
	if len(b) < 2 {
		return fmt.Errorf("NaN")
	}
	alph.Amount = new(big.Int)
	err := alph.Amount.UnmarshalJSON(b[1 : len(b)-1])
	return err
}

func ToNanoALPH(alph ALPH) int {
	m := new(big.Int).Div(alph.Amount, CoinInNanoALPH)
	return int(m.Int64())
}


metrics.go
package main

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"net/http"
)

type metrics struct {
	transferRun          prometheus.Counter
	txAmount             prometheus.Counter
	addressTotalBalance  *prometheus.GaugeVec
	addressLockedBalance *prometheus.GaugeVec
	addressUtxos         *prometheus.GaugeVec
}

func initPrometheus(env envConfig, mux *http.ServeMux) *metrics {
	m := &metrics{}

	m.transferRun = promauto.NewCounter(prometheus.CounterOpts{
		Name:      "transfer_runs_count",
		Help:      "Number of transfer runs",
		Namespace: env.MetricsNamespace,
		Subsystem: env.MetricsSubsystem,
	})

	m.txAmount = promauto.NewCounter(prometheus.CounterOpts{
		Name:      "transfer_amount_total",
		Help:      "Amount transferred",
		Namespace: env.MetricsNamespace,
		Subsystem: env.MetricsSubsystem,
	})

	m.addressTotalBalance = promauto.NewGaugeVec(prometheus.GaugeOpts{
		Name:      "total_balance",
		Help:      "Total balance of the address",
		Namespace: env.MetricsNamespace,
		Subsystem: env.MetricsSubsystem,
	}, []string{"address"})

	m.addressLockedBalance = promauto.NewGaugeVec(prometheus.GaugeOpts{
		Name:      "locked_balance",
		Help:      "Locked balance of the address",
		Namespace: env.MetricsNamespace,
		Subsystem: env.MetricsSubsystem,
	}, []string{"address"})

	m.addressUtxos = promauto.NewGaugeVec(prometheus.GaugeOpts{
		Name:      "utxos",
		Help:      "Number of UTXOs of the address",
		Namespace: env.MetricsNamespace,
		Subsystem: env.MetricsSubsystem,
	}, []string{"address"})

	mux.Handle(env.MetricsPath, promhttp.Handler())
	return m
}


transfer_handler.go
package main

import (
	"context"
	"fmt"
	alephium "github.com/alephium/go-sdk"
	"github.com/sirupsen/logrus"
	"sync"
	"time"
)

type transferHandler struct {
	alephiumClient     *alephium.APIClient
	walletName         string
	walletPassword     string
	mnemonicPassphrase string
	transferAddress    string
	transferMinAmount  ALPH
	transferFrequency  time.Duration
	immediate          bool
	metrics            *metrics
	log                *logrus.Logger
	concurrentExecLock *sync.RWMutex
}

func newTransferHandler(alephiumClient *alephium.APIClient, walletName string, walletPassword string,
	mnemonicPassphrase string, transferAddress string, transferMinAmount string, transferFrequency time.Duration,
	immediate bool, metrics *metrics, log *logrus.Logger) (*transferHandler, error) {

	minAlf, ok := ALPHFromCoinString(transferMinAmount)
	if !ok {
		return nil, fmt.Errorf("transferMinAmount %s is not a valid ALPH transfer amoount", transferMinAmount)
	}

	handler := &transferHandler{
		alephiumClient:     alephiumClient,
		walletName:         walletName,
		walletPassword:     walletPassword,
		mnemonicPassphrase: mnemonicPassphrase,
		transferAddress:    transferAddress,
		transferMinAmount:  minAlf,
		transferFrequency:  transferFrequency,
		immediate:          immediate,
		metrics:            metrics,
		log:                log,
		concurrentExecLock: &sync.RWMutex{},
	}

	return handler, nil
}

func (h *transferHandler) handle(ctx context.Context, log *logrus.Entry) error {
	if h.immediate {
		err := h.transfer(ctx, log)
		if err != nil {
			h.log.Debugf("Got an error while immediately transferring some amount. Err = %v", err)
			return err
		}
	}
	for range time.Tick(h.transferFrequency) {
		err := h.transfer(ctx, log)
		if err != nil {
			h.log.Debugf("Got an error while transferring some amount. Err = %v", err)
			return err
		}
	}
	return nil
}

func (h *transferHandler) transfer(ctx context.Context, log *logrus.Entry) error {

	locked := h.concurrentExecLock.TryLock()
	if !locked {
		log.Warnf("Another transfer process is still running...")
		return nil
	}
	defer h.concurrentExecLock.Unlock()

	h.metrics.transferRun.Inc()

	wallet, err := getWalletStatus(ctx, h.alephiumClient, h.walletName, log)
	if err != nil {
		h.log.WithError(err).Debugf("Got an error calling wallet status, err = %v", err)
		return err
	}
	if wallet.Locked {
		err := unlockWallet(ctx, h.alephiumClient, wallet.WalletName, h.walletPassword, h.mnemonicPassphrase, log)
		if err != nil {
			h.log.WithError(err).Debugf("Got an error calling wallet unlock. Err = %v", err)
			return err
		}
	}

	sweep := alephium.NewSweep(h.transferAddress)
	sweepAllReq := h.alephiumClient.WalletsApi.PostWalletsWalletNameSweepAllAddresses(ctx, wallet.WalletName).Sweep(*sweep)
	transferRes, _, err := sweepAllReq.Execute()
	if err != nil {
		h.log.WithError(err).Debugf("Got an error while sweeping all")
		return err
	}

	for _, tx := range transferRes.GetResults() {
		h.log.Infof("New tx %s,%d->%d just submitted", tx.TxId, tx.FromGroup, tx.ToGroup)
		var txConfirmed *alephium.Confirmed
		for txConfirmed == nil {

			txStatusReq := h.alephiumClient.TransactionsApi.GetTransactionsStatus(ctx)
			txStatus, _, err := txStatusReq.TxId(tx.TxId).ToGroup(tx.FromGroup).ToGroup(tx.ToGroup).Execute()
			if err != nil {
				h.log.WithError(err).Debugf("Got an error while getting tx status for tx %s", tx.TxId)
				return err
			}
			txConfirmed = txStatus.Confirmed
			time.Sleep(5 * time.Second)
		}
		h.log.Infof("New tx %s,%d->%d is now included in block %s!", tx.TxId, tx.FromGroup, tx.ToGroup, txConfirmed.BlockHash)
		// TODO: get block to find out the exact amount transferred to update counter consistently
	}

	return nil
}


address_balance_stats.go
package main

import (
	"context"
	alephium "github.com/alephium/go-sdk"
	"github.com/prometheus/client_golang/prometheus"
	"time"
)

type AddressBalanceStats struct {
	alephiumClient *alephium.APIClient
	addresses      []string
	metrics        *metrics
}

func newAddressBalanceStats(alephiumClient *alephium.APIClient, addresses []string, metrics *metrics) (*AddressBalanceStats, error) {
	handler := &AddressBalanceStats{
		alephiumClient: alephiumClient,
		addresses:      addresses,
		metrics:        metrics,
	}
	return handler, nil
}

func (h *AddressBalanceStats) Stats(ctx context.Context) error {
	err := h.doStats(ctx)
	if err != nil {
		return err
	}
	for range time.Tick(time.Minute) {
		select {
		case <-ctx.Done():
			return nil
		default:
		}
		err = h.doStats(ctx)
		if err != nil {
			return err
		}
	}
	return nil
}

func (h *AddressBalanceStats) doStats(ctx context.Context) error {
	for _, address := range h.addresses {
		addressBalanceReq := h.alephiumClient.AddressesApi.GetAddressesAddressBalance(ctx, address)
		balance, _, err := addressBalanceReq.Execute()
		if err != nil {
			return err
		}
		if addressBalance, ok := ALPHFromCoinString(balance.Balance); ok {
			h.metrics.addressTotalBalance.With(prometheus.Labels{"address": address}).Set(addressBalance.FloatALPH())
		}
		if addressLockedBalance, ok := ALPHFromCoinString(balance.LockedBalance); ok {
			h.metrics.addressLockedBalance.With(prometheus.Labels{"address": address}).Set(addressLockedBalance.FloatALPH())
		}
		h.metrics.addressUtxos.With(prometheus.Labels{"address": address}).Set(float64(balance.UtxoNum))
	}
	return nil
}


main.go
package main

import (
	"context"
	"flag"
	"fmt"
	alephium "github.com/alephium/go-sdk"
	"github.com/kelseyhightower/envconfig"
	"github.com/sirupsen/logrus"
	"github.com/sqooba/go-common/healthchecks"
	"github.com/sqooba/go-common/logging"
	"github.com/sqooba/go-common/version"
	"golang.org/x/sync/errgroup"
	"math/rand"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"
)

var (
	healthCheck = flag.Bool("health-check", false, "Run health-check")
	setLogLevel = flag.String("set-log-level", "", "Change log level. Possible values are trace,debug,info,warn,error,fatal,panic")
	log         = logging.NewLogger()
)

type envConfig struct {
	Port     string `envconfig:"PORT" default:"8080"`
	LogLevel string `envconfig:"LOG_LEVEL" default:"debug"`

	AlephiumEndpoint         string        `envconfig:"ALEPHIUM_ENDPOINT" default:"http://alephium:12973"`
	AlephiumApiKey           string        `envconfig:"ALEPHIUM_API_KEY" default:""`
	WalletName               string        `envconfig:"WALLET_NAME" default:"mining-companion-wallet-1"`
	WalletPassword           string        `envconfig:"WALLET_PASSWORD" default:"Default-Password-1234"`
	WalletMnemonic           string        `envconfig:"WALLET_MNEMONIC" default:""`
	WalletMnemonicPassphrase string        `envconfig:"WALLET_MNEMONIC_PASSPHRASE" default:""`
	TransferMinAmount        string        `envconfig:"TRANSFER_MIN_AMOUNT" default:"20000000000000000000"`
	TransferAddress          string        `envconfig:"TRANSFER_ADDRESS" default:""`
	TransferFrequency        time.Duration `envconfig:"TRANSFER_FREQUENCY" default:"15m"`
	PrintMnemonic            bool          `envconfig:"PRINT_MNEMONIC" default:"false"`
	ImmediateTransfer        bool          `envconfig:"IMMEDIATE_TRANSFER" default:"false"`

	MetricsNamespace string `envconfig:"METRICS_NAMESPACE" default:"alephium"`
	MetricsSubsystem string `envconfig:"METRICS_SUBSYSTEM" default:"miningcompanion"`
	MetricsPath      string `envconfig:"METRICS_PATH" default:"/metrics"`
}

const (
	DefaultWalletPassword = "Default-Password-1234"
)

func main() {

	log.Println("Alephium-mining-companion application is initializing...")
	log.Printf("Version    : %s", version.Version)
	log.Printf("Commit     : %s", version.GitCommit)
	log.Printf("Build date : %s", version.BuildDate)
	log.Printf("OSarch     : %s", version.OsArch)

	rand.Seed(time.Now().UnixNano())

	var env envConfig
	if err := envconfig.Process("", &env); err != nil {
		log.Fatalf("Failed to process env var: %s\n", err)
		return
	}

	flag.Parse()

	err := logging.SetLogLevel(log, env.LogLevel)
	if err != nil {
		log.Fatalf("Logging level %s do not seem to be right. Err = %v", env.LogLevel, err)
	}

	// Running health check (so that it can be the same binary in the containers
	if *healthCheck {
		healthchecks.RunHealthCheckAndExit(env.Port)
	}
	if *setLogLevel != "" {
		logging.SetRemoteLogLevelAndExit(log, env.Port, *setLogLevel)
	}

	if env.WalletName == "" || env.WalletPassword == "" {
		log.Fatalf("Some mandatory configuration parameters are missing. Please correct the config and retry.")
	}
	if env.WalletPassword == DefaultWalletPassword {
		log.Warnf("Your using the default password. This is not recommanded for production use.")
	}

	// Register health checks and metrics
	initHealthChecks(env, http.DefaultServeMux)
	metrics := initPrometheus(env, http.DefaultServeMux)

	// Special endpoint to change the verbosity at runtime, i.e. curl -X PUT --data debug ...
	logging.InitVerbosityHandler(log, http.DefaultServeMux)

	// errgroup will coordinate the many routines handling the API.
	cancellableCtx, cancel := context.WithCancel(context.Background())
	g, ctx := errgroup.WithContext(cancellableCtx)
	signalChan := make(chan os.Signal, 1)
	signal.Notify(signalChan, syscall.SIGINT, syscall.SIGTERM)

	s := http.Server{Addr: fmt.Sprint(":", env.Port)}
	g.Go(s.ListenAndServe)

	alephiumConfig := alephium.NewConfiguration()
	alephiumConfig.Host = env.AlephiumEndpoint
	if log.Level >= logrus.TraceLevel {
		alephiumConfig.Debug = true
	}
	if env.AlephiumApiKey != "" {
		alephiumConfig.DefaultHeader["X-API-KEY"] = env.AlephiumApiKey
	}
	alephiumClient := alephium.NewAPIClient(alephiumConfig)
	if err != nil {
		log.WithError(err).Fatalf("Got an error instantiating alephium client on %s", env.AlephiumEndpoint)
	}

	miningHandler, err := newMiningHandler(alephiumClient, env.WalletName, env.WalletPassword,
		env.WalletMnemonic, env.WalletMnemonicPassphrase, env.PrintMnemonic, log)
	if err != nil {
		log.Fatalf("Got an error while creating the wallet handler. Err = %v", err)
	}

	wallet, err := miningHandler.createAndUnlockWallet(ctx, logrus.NewEntry(log))
	if err != nil {
		log.Fatalf("Got an error while creating and/or unlocking the wallet %s. Err = %v", env.WalletName, err)
	}

	err = miningHandler.updateMinersAddresses(ctx, logrus.NewEntry(log))
	if err != nil {
		log.Fatalf("Got an error while updating miners addresses. Err = %v", err)
	}

	minersAddressesReq := alephiumClient.MinersApi.GetMinersAddresses(ctx)
	minersAddresses, _, err := minersAddressesReq.Execute()
	if err != nil {
		log.WithError(err).Fatalf("Got an error calling miners addresses")
	}
	log.Infof("Mining wallet %s (with addresses %v) is ready to be used, now waiting for the node to become in sync if needed.",
		wallet.WalletName, minersAddresses.Addresses)

	err = miningHandler.waitForNodeInSync(ctx, logrus.NewEntry(log))
	if err != nil {
		log.Fatalf("Got an error while waiting for the node to be in sync with peers. Err = %v", err)
	}
	g.Go(func() error {
		return miningHandler.ensureMiningWalletAndNodeMining(ctx, logrus.NewEntry(log))
	})

	addressesToWatch := make([]string, 0, len(minersAddresses.Addresses)+1)
	for _, a := range minersAddresses.Addresses {
		addressesToWatch = append(addressesToWatch, a)
	}
	if env.TransferAddress != "" {
		addressesToWatch = append(addressesToWatch, env.TransferAddress)
	}
	addressBalanceStats, _ := newAddressBalanceStats(alephiumClient, addressesToWatch, metrics)
	g.Go(func() error { return addressBalanceStats.Stats(ctx) })

	if env.TransferAddress != "" {
		transferHandler, err := newTransferHandler(alephiumClient, wallet.WalletName, env.WalletPassword,
			env.WalletMnemonicPassphrase, env.TransferAddress, env.TransferMinAmount, env.TransferFrequency,
			env.ImmediateTransfer, metrics, log)
		if err != nil {
			log.WithError(err).Fatalf("Got an error while instanciating the transfer handler")
		}

		log.Infof("We will transfer to %s the mining reward every %s.", env.TransferAddress, env.TransferFrequency)

		g.Go(func() error {
			err := transferHandler.handle(ctx, logrus.NewEntry(log))
			if err != nil {
				log.WithError(err).Fatalf("Got an error while handling the transfer")
			}
			return err
		})
	} else {
		log.Infof("No transfer address configure, no problem, job is done.")
		cancel()
	}

	// Wait for any shutdown
	select {
	case <-signalChan:
		log.Info("Shutdown signal received, exiting...")
		cancel()
		break
	case <-ctx.Done():
		log.Info("Group context is done, exiting...")
		cancel()
		break
	}

	err = ctx.Err()
	if err != nil {
		log.WithError(err).Fatal("Got an error from the error group context")
	}

	log.Infof("All good, stopping now.")
}
